### PR TITLE
urlapi: stricter CURLUPART_PORT parsing

### PR DESCRIPTION
**Pull request by bagder**
Only allow well formed decimal numbers in the input.

Document that the number MUST be between 1 and 65535.

Add tests to test 1560 to verify the above.

Ref: https://github.com/curl/curl/issues/3753

### CODE REVIEW

File: lib/urlapi.c [hunk 1]
```diff
@@ -1188,12 +1188,18 @@ CURLUcode curl_url_set(CURLU *u, CURLUPart what,
...
+  {
+    char *endp;
     urlencode = FALSE; /* never */
-    port = strtol(part, NULL, 10);  /* Port number must be decimal */
+    port = strtol(part, &endp, 10);  /* Port number must be decimal */
     if((port <= 0) || (port > 0xffff))
       return CURLUE_BAD_PORT_NUMBER;
+    if(*endp)
+      /* weirdly provided number, not good! */
+      return CURLUE_BAD_PORT_NUMBER;
```
Comment (danielgustafsson) [inline lib/urlapi.c]: Isn't this better classified with `CURLUE_MALFORMED_INPUT`?

Comment (bagder) [reply lib/urlapi.c]: I agree. Will amend!

**Comment by bagder** [comment]
Nice fix @jay. I think we should also add a test for that case. I can do that tomorrow.

**Comment by jay** [comment]
I might not have gone far enough. The reporter in #3753 basically showed that CURLOPT_PORT 1234L set that port in his `curlu` but CURLOPT_PORT 0L didn't reset it to the default port. Of course neither one of those things should have had any effect on his `curlu` and you fixed that in #3761. Separately from that however is still the issue of how it is treated internally. The reporter repro boiled down is basically this:

~~~
curl_easy_setopt(curl, CURLOPT_CURLU, curlu);
curl_easy_setopt(curl, CURLOPT_PORT, 1234L);
curl_code = curl_easy_perform(curl);
// now reset port to go back to port 80 (not!)
curl_easy_setopt(curl, CURLOPT_PORT, 0L);
curl_code = curl_easy_perform(curl);
~~~

Internally `parse_remote_port()` will be called in the first case to update the internal CURLU (ie the CURLU that would be copied from the user's `curlu` in this example). Yet that does not happen in the second case since `data->set.use_port` is 0:

https://github.com/curl/curl/blob/521bbbe29928f9bc1c61306df612e856d45cbe5a/lib/url.c#L3009-L3025

My fix would allow the user to use `curl_url_set(uh, CURLUPART_PORT, "", 0);` to reset the port in their own CURLU (which is read-only for us) but doesn't address this with the internal CURLU. In other words the user still can't use CURLOPT_PORT 0L to reset the port in the URL AFAICS (granted I haven't actually tried this yet).

**Comment by bagder** [comment]
> I might not have gone far enough.

I think you're overthinking it. This PR and change is only about how to parse the `CURLUPART_PORT` string when `curl_url_set` is called.

The #3753 issue was totally separate. It occurred primarily because curl modified the handle that it wasn't supposed to touch so even when setting `CURLOPT_PORT` to zero, the port number in the URL wasn't correctly reset. That part was fixed in #3761.

**Comment by bagder** [comment]
Oh sorry, this extra `""` support is unnecessary. The man page actually already says:

> Setting a part to a NULL pointer will effectively remove that part's contents from the CURLU handle.

So NULL, not "", has that effect. I'll add a test case for it.

**Comment by jay** [comment]
Oops. I also missed that in the code and now it seems totally obvious. Does portnum need to be cleared:

~~~diff
diff --git a/lib/urlapi.c b/lib/urlapi.c
index 3da8ff8..0eb06d2 100644
--- a/lib/urlapi.c
+++ b/lib/urlapi.c
@@ -1145,6 +1145,7 @@ CURLUcode curl_url_set(CURLU *u, CURLUPart what,
       storep = &u->host;
       break;
     case CURLUPART_PORT:
+      u->portnum = 0;
       storep = &u->port;
       break;
     case CURLUPART_PATH:
~~~

**Comment by bagder** [comment]
yes!

**Pull request closed without merging**